### PR TITLE
implement G-UNI interactive adapter

### PR DESCRIPTION
**Pull request by superarius**
G-UNI tokens are single LP positions on Uniswap V3 token pairs that are fractionalized and thus made fungible. Anyone can add or remove their liquidity to this shared position and is entitled to their portion of the fees earned. Gelato bot network can call a method to automatically reinvest any fees earned by the position (only once enough fees have accured) to create compounding effect. The position range of any G-UNI position is static and cannot be changed by the gelato bots (however a permissioned 'manager' role can be assigned on pools and this manager has the power to adjust the range with an executiveRebalance() call)

one note: I basically followed as closely as possible the patterns of the UniswapV2AssetInteractiveAdapter, which doesn't use the uniswap router but mints and burns directly on the pool. So we did the same thing with G-UNI (did not use guni router but mint/burn directly on the pool) but this way things like WETH/ETH support or swapping some inventory of user inputs before deposit (so that they can deposit maximal liquidity) must be handled natively.

### CODE REVIEW

**Review by sobolev-igor — CHANGES_REQUESTED** [review]

**Comment by sobolev-igor** [comment]
Hey @superarius I've fixed tests, now `nom run test:interactiveAdapters` runs the tests and they pass successfully!
I've made some fix suggestions, but they are mostly code-style level.

File: contracts/interfaces/GUniPool.sol [hunk 1]
```diff
@@ -0,0 +1,47 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity 0.7.3;
+
+interface GUniPool {
```
[the remark this answers is not in the document]
Comment (superarius) [reply contracts/interfaces/GUniPool.sol]: done

File: contracts/interactiveAdapters/gelato/GUniInteractiveAdapter.sol [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+        address token = tokenAmounts[0].token;
+        uint256 amount = getAbsoluteAmountWithdraw(tokenAmounts[0]);
+
+        // solhint-disable-next-line no-empty-blocks
+        try GUniPool(token).burn(amount, address(this)) returns (
+            uint256 amount0,
+            uint256 amount1,
+            uint128
+        ) {
```
[the remark this answers is not in the document]
Comment (superarius) [reply contracts/interactiveAdapters/gelato/GUniInteractiveAdapter.sol:103]: should be impossible but added anyway as sanity check

File: contracts/interactiveAdapters/gelato/GUniInteractiveAdapter.sol [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+     */
+    function withdraw(TokenAmount[] calldata tokenAmounts, bytes calldata)
+        external
+        payable
+        override
+        returns (address[] memory tokensToBeWithdrawn)
+    {
+        require(tokenAmounts.length == 1, "GUIA: should be 1 tokenAmount");
+
+        address token = tokenAmounts[0].token;
```
[the remark this answers is not in the document]
Comment (superarius) [reply contracts/interactiveAdapters/gelato/GUniInteractiveAdapter.sol]: done

File: contracts/interactiveAdapters/gelato/GUniInteractiveAdapter.sol [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+/**
+ * @title Interactive adapter for G-UNI protocol (fungible lp tokens on Uniswap v3).
+ * @dev Implementation of InteractiveAdapter abstract contract.
+ * @author Igor Sobolev <sobolev@zerion.io>
+ */
+contract GUniInteractiveAdapter is InteractiveAdapter, ERC20ProtocolAdapter {
+    using SafeERC20 for ERC20;
+
+    address internal constant WETH = 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2;
```
[the remark this answers is not in the document]
Comment (superarius) [reply contracts/interactiveAdapters/gelato/GUniInteractiveAdapter.sol]: removed

File: contracts/interactiveAdapters/gelato/GUniInteractiveAdapter.sol [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+                tokensToBeWithdrawn = new address[](1);
+                tokensToBeWithdrawn[0] = GUniPool(token).token0();
+            } else {
+                tokensToBeWithdrawn = new address[](1);
+                tokensToBeWithdrawn[0] = GUniPool(token).token1();
+            }
+        } catch Error(string memory reason) {
+            revert(reason);
+        } catch {
+            revert("ULIA: withdraw fail");
```
[the remark this answers is not in the document]
Comment (superarius) [reply contracts/interactiveAdapters/gelato/GUniInteractiveAdapter.sol]: done in PR

File: contracts/interactiveAdapters/gelato/GUniInteractiveAdapter.sol [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+import { SafeERC20 } from "../../shared/SafeERC20.sol";
+import { TokenAmount } from "../../shared/Structs.sol";
+import { ERC20ProtocolAdapter } from "../../adapters/ERC20ProtocolAdapter.sol";
+import { InteractiveAdapter } from "../InteractiveAdapter.sol";
+import { GUniPool } from "../../interfaces/GUniPool.sol";
+
+/**
+ * @title Interactive adapter for G-UNI protocol (fungible lp tokens on Uniswap v3).
+ * @dev Implementation of InteractiveAdapter abstract contract.
+ * @author Igor Sobolev <sobolev@zerion.io>
```
[the remark this answers is not in the document]
Comment (superarius) [reply contracts/interactiveAdapters/gelato/GUniInteractiveAdapter.sol]: 👍

**Review by sobolev-igor — APPROVED** [review]

**Pull request merged**